app-backend/tile/src/main/scala/image/Mosaic.scala

```diff
@@ -36,123 +38,88 @@ object Mosaic {
       } catch {
         case e: AttributeNotFoundError if tryZoom > 0 => readMetadata(store, tryZoom - 1)
       }
 
-    tileLayerMetadataCache.get(s"mosaic-tlm-$id-$zoom", { cKey =>
-      for {
-        prefix <- LayerCache.prefixFromLayerId(id)
-        store <- LayerCache.attributeStore(prefix)
-      } yield {
+    memcached.caching(s"mosaic-tlm-$id-$zoom") { _ =>
+      OptionT(LayerCache.attributeStoreForLayer(id)).mapFilter { store =>
         readMetadata(store, zoom)
-      }
-    })
+      }.value
+    }
   }
 
-
-  val mosaicDefinitionCache = HeapBackedMemcachedClient[Option[MosaicDefinition]](memcachedClient)
-  def mosaicDefinition(projectId: UUID, tagttl: Option[TagWithTTL])(implicit db: Database) = {
+  def mosaicDefinition(projectId: UUID, tagttl: Option[TagWithTTL])(implicit db: Database): Future[Option[MosaicDefinition]] = {
+    // TODO: TTL in tagttl does nothing, default memcachd TTL is used anyway
     val cacheKey = tagttl match {
       case Some(t) => s"mosaic-definition-$projectId-${t.tag}"
       case None => s"mosaic-definition-$projectId"
     }
-    mosaicDefinitionCache.caching(cacheKey) { implicit ec =>
+
+    memcached.caching(cacheKey) { _ =>
       ScenesToProjects.getMosaicDefinition(projectId)
     }
   }
 
   /** Fetch the tile for given resolution. If it is not present, use a tile from a lower zoom level */
-  def fetch(id: UUID, zoom: Int, col: Int, row: Int)(implicit database: Database): Future[Option[MultibandTile]] =
-    tileLayerMetadata(id, zoom).flatMap {
-      case Some((sourceZoom, tlm)) =>
-        val zdiff = zoom - sourceZoom
-        val pdiff = 1 << zdiff
-        val sourceKey = SpatialKey(col / pdiff, row / pdiff)
-        if (tlm.bounds includes sourceKey)
-          for ( maybeTile <- LayerCache.maybeTile(id, sourceZoom, sourceKey) ) yield {
-            for (tile <- maybeTile) yield {
-              val innerCol  = col % pdiff
-              val innerRow  = row % pdiff
-              val cols = tile.cols / pdiff
-              val rows = tile.rows / pdiff
-              tile.crop(GridBounds(
-                colMin = innerCol * cols,
-                rowMin = innerRow * rows,
-                colMax = (innerCol + 1) * cols - 1,
-                rowMax = (innerRow + 1) * rows - 1
-              )).resample(256, 256)
-            }
-          }
-        else
-          Future.successful(None)
-      case None =>
-        Future.successful(None)
+  def fetch(id: UUID, zoom: Int, col: Int, row: Int)(implicit database: Database): Future[Option[MultibandTile]] = {
+    for {
+      (sourceZoom, tlm) <- OptionT(tileLayerMetadata(id, zoom))
+      zoomDiff = zoom - sourceZoom
+      resolutionDiff = 1 << zoomDiff
+      sourceKey = SpatialKey(col / resolutionDiff, row / resolutionDiff)
+      tile <- OptionT(LayerCache.layerTile(id, sourceZoom, sourceKey)) if tlm.bounds.includes(sourceKey)
```

Comment:
Monad Transformer is just a mechanism to compose two things that have `map` or `flapMap`. It turns this:

```scala
def foo: Future[Option[A]]
def bar: A => Future[Option[B]]
def baz: B => Future[Option[C]]

foo.flatMap({
  case None => ???  // deal with it
  case Some(a) => bar(a).flatMap({
    case None => ??? // deal with it again
    case Some(b) => baz(b)
  })
})
```
into this:
```scala
foo.flatMap(bar).flatMap(baz)
```